### PR TITLE
Enhance source line finding logic for debuginfo

### PR DESCRIPTION
**Pull request by sklam**
Enhance source line finding logic introduced in https://github.com/numba/numba/pull/7744/files. This enables correct reporting of source linenos for functions containing unusually indented comments.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
@sklam given the accidental addition of a notebook removed in cf2047e, perhaps rewrite and flatten history to a single patch?

**Comment by sklam** [comment]
Squashed all commits as suggested

**Comment by sklam** [comment]
Oh no, i messed it up. There's a main merge that i squashed

**Comment by sklam** [comment]
It's ok now

**Comment by stuartarchibald** [comment]
Should this sort of thing be able to work:
```python
from contextlib import ContextDecorator
from numba import njit
from numba.misc.firstlinefinder import get_func_body_first_lineno

class disruptive(ContextDecorator):
    pass

@disruptive()
@njit
def foo():
    pass

print(get_func_body_first_lineno(foo))
```
currently gives line 78!?

**Comment by sklam** [comment]
78?!

**Comment by sklam** [comment]
Oh, i need to add the requirement that the `foo.__code__` object is the one being tracked. The `disruptive` is making `foo.__code__` points to `contextlib.py`.

File: numba/tests/test_firstlinefinder.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+from numba import njit
+from numba.tests.support import TestCase
+from numba.misc.firstlinefinder import get_func_body_first_lineno
+
+
+class TestFirstLineFinder(TestCase):
+    """
+    The following methods contains tests that are sensitive to the source
+    locations w.r.t. the begining of each method.
```
Comment (stuartarchibald) [inline numba/tests/test_firstlinefinder.py:12]:
```suggestion
    locations w.r.t. the beginning of each method.
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, few minor things to look at but otherwise looks good.

File: numba/tests/test_debuginfo.py [hunk 2]
```diff
@@ -293,9 +293,10 @@ def foo(a):
                 groups = matched.groups()
                 self.assertEqual(len(groups), 1)
                 dbg_line = int(groups[0])
-                # +1 for the decorator on Python 3.8+, `inspect` changed, also
-                # recall that Numba's DWARF refers to the "def" line
-                defline = pysrc_line_start + (utils.PYVERSION >= (3, 8))
+                # +1 for the decorator.
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_debuginfo.py:296]: Yes, the use of AST means the Numba DWARF info will always be correctly pointing to the `def` line.

File: numba/misc/firstlinefinder.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import ast
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/misc/firstlinefinder.py:6]: done in https://github.com/numba/numba/pull/7805/commits/dbf2e47211ed2fcdf6fbaec2e18fc9fb5394dc52

File: numba/tests/test_debuginfo.py [hunk 4]
```diff
@@ -753,14 +754,13 @@ def foo():
...
 
-        self.assertEqual(len(w), 1)
-        found = w[0]
-        self.assertEqual(found.category, NumbaDebugInfoWarning)
-        msg = str(found.message)
-        # make sure the warning contains the right message
-        self.assertIn('Could not parse the source for function', msg)
-        # and refers to the offending function
-        self.assertIn(str(foo.py_func), msg)
+        # No warnings
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_debuginfo.py:757]: done in https://github.com/numba/numba/pull/7805/commits/e9464ce0570e47fe791436420eb362a1c47763d2

File: numba/misc/firstlinefinder.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                # Yes, we found the function.
+                # So, use the first statement line as the first line.
+                if node.body:
+                    first_stmt = node.body[0]
+                    self.first_stmt_line = first_stmt.lineno
+                    return
+                else:
+                    # This is probably unreachable.
+                    # Function body cannot be bare. It must at least have
+                    # A const string for docstring or a `pass`.
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/misc/firstlinefinder.py:62]: done in https://github.com/numba/numba/pull/7805/commits/f8de4ceee2caa07bf1d86d1d0e1b3ec86861717c

File: numba/misc/firstlinefinder.py [hunk 6]
```diff
@@ -1,5 +1,12 @@
+"""
+This module provides helper functions to find the first line of the function
+body.
```
Comment (stuartarchibald) [inline numba/misc/firstlinefinder.py:3]:
```suggestion
This module provides helper functions to find the first line of a function
body.
```

File: numba/tests/test_debuginfo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_debuginfo.py:296]: Great, thanks!

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes, a one word change to resolve else looks good. Many thanks.

**Pull request merged**